pennylane/gradients/spsa_gradient.py

```diff
@@ -0,0 +1,573 @@
+          (array(2.548), array(2.548), array(2.548))),
+         ((array(-1.06), array(-1.06), array(-1.06)),
+          (array(1.90588), array(1.90588), array(1.90588))))
+
+        The outermost tuple contains results corresponding to each element of the shot vector,
+        as is also visible by the increasing precision.
+        Note that the stochastic approximation and the fluctuations from the shot noise
+        of the device accumulate, leading to a very coarse-grained estimate for the gradient.
+    """
+    if argnum is None and not tape.trainable_params:
+        return _no_trainable_grad_new(tape, shots)
+
+    if validate_params:
+        if "grad_method" not in tape._par_info[0]:
+            gradient_analysis(tape, grad_fn=_spsa_grad_new)
+        diff_methods = grad_method_validation("numeric", tape)
+    else:
+        diff_methods = ["F" for i in tape.trainable_params]
+
+    if all(g == "0" for g in diff_methods):
+        return _all_zero_grad_new(tape, shots)
+
+    gradient_tapes = []
+    extract_r0 = False
+
+    coeffs, shifts = finite_diff_coeffs(n=n, approx_order=approx_order, strategy=strategy)
+
+    if 0 in shifts:
+        # Finite difference formula includes a term with zero shift.
+
+        if f0 is None:
+            # Ensure that the unshifted tape is appended to the gradient tapes
+            gradient_tapes.append(tape)
+            extract_r0 = True
+
+        # Skip the unshifted tape
+        shifts = shifts[1:]
+
+    method_map = choose_grad_methods(diff_methods, argnum)
+
+    indices = [
+        i for i, _ in enumerate(tape.trainable_params) if (i in method_map and method_map[i] != "0")
+    ]
+
+    tapes_per_grad = len(shifts)
+    all_coeffs = []
+    for idx_rep in range(num_directions):
+        direction = sampler(indices, len(tape.trainable_params), idx_rep)
+        inv_direction = qml.math.divide(
+            1, direction, where=(direction != 0), out=qml.math.zeros_like(direction)
+        )
+        # Use only the non-zero part of `direction` for the shifts, to skip redundant zero shifts
+        _shifts = qml.math.tensordot(h * shifts, direction[indices], axes=0)
+        all_coeffs.append(qml.math.tensordot(coeffs / h**n, inv_direction, axes=0))
+        g_tapes = generate_multishifted_tapes(tape, indices, _shifts)
+        gradient_tapes.extend(g_tapes)
+
+    def _single_shot_batch_result(results):
+        """Auxiliary function for post-processing one batch of results corresponding to finite
+        shots or a single component of a shot vector"""
+
+        r0, results = (results[0], results[1:]) if extract_r0 else (f0, results)
+        if len(tape.measurements) == 1:
+            grads = 0
+            for rep, _coeffs in enumerate(all_coeffs):
+                res = results[rep * tapes_per_grad : (rep + 1) * tapes_per_grad]
+                if r0 is not None:
+                    res.insert(0, r0)
+                res = qml.math.stack(res)
```

Comment:
I am worried with stacking here. Is it possible to have maybe different shaped arrays at this point? It would trigger a ragged array warning.

Reply:
No, all results here should have the same shape, because they  only differ by the shifted _input_ values, and we are in the single-measurement case. But I might be missing some edge case here? Happy to discuss this further! :)

Reply:
No I think you are correct!